### DISCUSSION TITLE
Changes to Build Process Have Made Development Difficult

### DISCUSSION THREAD

**Discussion by ChristopherDavenport** [http4s/http4s#5709] [General]
- quicklint 90+ seconds
- Warns locally fail in CI which leads to multiple failure cycles independent of quicklint
- Only minimal changes get review cycles. Everything larger gets month cycles, and faster moving changes to those same components then make coming back to get them in line have massive diffs, leading to exhaustion trying to make the larger (even wanted) changes.
- Focus on sterility, loses innovation. If its small getting it great is one thing, if its big it's a massive undertaking. Larger changes would often be better to turn it all off, and then let later PR's lower it back down as paying down acquired tech debt rather than never taking tech debt initially.

Anything substantial that's not net new and small enough to fit into a 5 minute read I currently question whether its efficient to PR, since the build frustrates, reviews linger, and there are other mechanisms to offer code.

This didn't use to be the case, and I think there is likely a happy medium, but right now I am struggling with it. Facing this build is no longer about delivering value. Working on the code is a joy, but facing the build takes all the fun out for me.

**Comment by ChristopherDavenport** [comment]
I've spent 5 hours tonight, and barely made a dent. I've cried 3 times tonight because a year of my life is locked behind a puzzle that I don't know how to solve. 

A brand new feature has had zero comments or thoughts besides for @armanbilge leading me to ask why am I doing this if none of the other maintainers are even invested in any of the things I'm trying to contribute. I didn't need a review, but silence over an entire new type of http for a whole week.

**Reply by danicheg** [reply]
You are definitely doing great stuff, @ChristopherDavenport.
Personally speaking, I don't rush into review for "Draft" PRs. Just because things may be changed dramatically (as it happens from my experience). Also, your PR has a really vast diff, so some notes for it would be pretty helpful in a review. But yeah, I exactly will spend N hours trying to suss it out your PR :)

**Reply by rossabaker** [reply]
I deleted a reply that was too self-indulgent.  I will try again:

1. I'm sincerely sorry for the frustration, but please remember that this is a challenging time of year for many people.  Capacities ebb and flow even where interest is high.  
2. To wit, it took me five days to even see this conversation.
3. Echoing @danicheg's point, pointers to the crucial parts of the PR would make it less daunting and may expedite the turnaround.

**Comment by armanbilge** [comment]
> Only minimal changes get review cycles. Everything larger gets month cycles, and faster moving changes to those same components then make coming back to get them in line have massive diffs, leading to exhaustion trying to make the larger (even wanted) changes.

I'm too new to know or begin to appreciate all the history of this project, but as it stands today I wonder if the ratio of active maintainer time to the size and complexity of this project is out of calibration, especially considering that this project supposed to be in development (and not just maintenance). I think the effort required simply to maintain three branches with 20+ modules (almost 30??) contributes to that stifling of innovation and slow cycle of reviews. I realize that I have personally contributed a significant amount of this complexity ...

I'm also reminded of the rather successful call for community maintainers in https://github.com/typelevel/general/issues/111. I'd like to imagine that explicitly dialing back the scope of the core project (and thus, the responsibilities of the core maintainers) along with a call-to-action can help to restore the balance @ChristopherDavenport is seeking while also mobilizing the community to step up for the stuff they care about.

As far as I understand, http4s 1.0 is headed in this direction anyway, except we are having a hard time reaching there at the current pace.

**Reply by bplommer** [reply]
> I think the effort required simply to maintain three branches with 20+ modules (almost 30??) contributes to that stifling of innovation and slow cycle of reviews

Agreed. Could be time to make 0.22 and maybe 0.23 bugfixes-only?

**Reply by rossabaker** [reply]
The Great Schism is difficult as long as prior branches are being maintained.  Merging from the monorepo to the polyrepos needs to be exceptional.  So how do we get there from here?

1. We could frost 0.23 and schism main.
    1. Do it now, and we deliver no incremental value until a rushed 1.0.
    2. Do it late, and we have increasingly fraught merges before a delayed 1.0.
2. We could frost 0.22 and schism series/0.23.  Then what?
    1. Publish non-core as 0.23.x on core 0.23.y.  Users are confused that x != y in the middle of 0.23.
    2. Publish non-core as 0.24.x on 0.23.y core.  Users are confused that 0.23 != 0.24 and why non-core modules are disappearing from 0.23.
    3. Publish non-core as 0.24.x on core 0.24.y.  Users are confused that x != y and wtf is 0.24, but at least we got their attention.

**Reply by armanbilge** [reply]
I think there is an option you left out, 2iv: publish non-core as 0.23.x on core 0.23.y, except keep x and y in sync. Yes, release day will be annoying as hell but for most of the repos it should be a simple bump and tag. This is what I'm currently doing with http4s-dom to track the 1.0.0 milestones because otherwise it's just too confusing.

**Reply by armanbilge** [reply]
It will be easier if we can recruit community maintainers to do the bump and tag :)

**Reply by armanbilge** [reply]
Also, I know how to automate the bump: we can create a repo with a workflow to manually run scala-steward on all http4s-org repos. This is super easy to setup with a https://github.com/scala-steward-org/scala-steward-action

So, on release day, as soon as core hits maven, we run our steward action. It'll open PRs on all the satellite repos. We slam merge and tag.

**Reply by armanbilge** [reply]
If we are _really_ clever, we can get mergify to merge those PRs automatically and maybe even do the tagging too somehow 🤔

**Reply by rossabaker** [reply]
For that to work, we have to be wary of source-breaking changes in core, or else the satellite modules might have trouble adapting in a binary-compatible way.  That's okay: source-breaking changes in a patch release should be exceptional anyway.  I like how that solution is not disruptive to users of 0.23, and mostly automatable.

There still needs to be a plan to merge from the monorepo to the polyrepos in case of CVEs or whatever other maintenance is deemed necessary pre-schism.  If the polyrepos share the full history, it's straight merges, throwing away any conflicts outside the relevant directories.  If the polyrepos use a history based on git filtering, we just need to filter origin with the same parameters and merge that.  Either of those might be assisted with a script called by a workflow.

**Reply by armanbilge** [reply]
> There still needs to be a plan to merge from the monorepo to the polyrepos in case of CVEs or whatever other maintenance is deemed necessary pre-schism.

Unpopular opinion (?): please, no more 0.22 -> 0.23 merges. Develop on 0.23, and cherry-pick/backport to 0.22 as necessary (hopefully only for CVEs, which we won't hope for). This is how CE and FS2 manage their 2.x branches.

Otherwise, those git things sound good. I say just keep the history since that sounds simpler.

**Reply by rossabaker** [reply]
> Develop on 0.23

This seems justifiable.  Last month's stats:

- 0.18.x: 2%
- 0.20.x: 3%
- 0.21.x: 47%
- 0.22.x: 9%
- 0.23.x: 23%
- 1.x: 14%

> cherry-pick/backport to 0.22 as necessary 

Disagree here.  When we know in advance a change is desired on two branches, merging it forward instead of cherry-picking it backward keeps the histories related and simplifies the next one.  The first point means fewer merges, but the current logistics make those fewer merges easier.

**Reply by armanbilge** [reply]
>  When we know in advance a change is desired on two branches, merging it forward instead of cherry-picking it backward keeps the histories related and simplifies the next one.

👍 sounds good.

So, to execute, we need N new repos 😅

**Reply by rossabaker** [reply]
I had started a servlet one, but I think it's off on the wrong foot.  When I get a moment (ha!), I'll try a couple strategies in my account and share.

**Comment by bplommer** [comment]
> quicklint 90+ seconds

Scalafix has an option to only run fixes against files that changed from the upstream branch - I had problems getting it working but I can take another crack at it.

> Warns locally fail in CI which leads to multiple failure cycles independent of quicklint

It should be possible to make warnings fatal when running quicklint 

> Only minimal changes get review cycles. Everything larger gets month cycles, and faster moving changes to those same components then make coming back to get them in line have massive diffs, leading to exhaustion trying to make the larger (even wanted) changes.

Unfortunately this is a problem with pull requests as a workflow in general IME. My only suggestion for making this better is to try switching to a workflow where the concept for a change is approved and then the code changes are done as a series of smaller PRs that can be merged incrementally. #4509 and #5251 are examples of PRs I'm excited about (or was a couple of months ago) where I think this could work.

Unfortunately this is affected by something I've long found frustrating with open source in general, which is that feedback on an idea is much less likely to be forthcoming if you just open an issue describing the idea, compared to if you put in the work of writing the code.

> - Focus on sterility, loses innovation. If its small getting it great is one thing, if its big it's a massive undertaking. Larger changes would often be better to turn it all off, and then let later PR's lower it back down as paying down acquired tech debt rather than never taking tech debt initially.

Agreed, at least for `main`

**Reply by armanbilge** [reply]
> My only suggestion for making this better is to try switching to a workflow where the concept for a change is approved and then the code changes are done as a series of smaller PRs that can be merged incrementally

💯 to this. Vasil used this technique to great effect to gradually implement fiber dumps in CE3.3 and it made it possible for me to review his PRs and give useful feedback, despite the fact it was tinkering with scary things like the work-stealing threadpool and the `IO` runloop. Essentially, this technique helps make your PR more accessible.

**Reply by bplommer** [reply]
You've inspired me to open #5712 as a minimum viable PR toward #5251  😄

**Reply by armanbilge** [reply]
So now I'm on the hook for a review, eh? 😝

**Reply by bplommer** [reply]
I would never suggest such a thing, especially given how long I've owed @m3is0 some reviews...

**Reply by rossabaker** [reply]
Linting and formatting sucks during exploratory phases, which is I think why warnings are non-fatal until CI in sbt-spiewak and inherted by sbt-http4s-org.  I also think they improve code quality and legibility, particularly on projects with many contributors and tens of thousands of lines of code.

Could we detect draft PRs and leave warnings non-fatal for early iteration on these big ideas, and enforce only when the PR is ready to review?
